keps/sig-node/3759-sidecar-containers/README.md

```diff
@@ -0,0 +1,1235 @@
+  OnFailure restartPolicy indicating the critical Initialization step.
+  Supporting this value on Pods with the the restartPolicy “Always” is not
+  planned as it changes the semantic of this policy and may affect how
+  deployments control the # of Pods in this deployment.
+- `OnFailure` indicates the init container will be restarted until it completes
+  successfully. This value may be used for Pods with the restartPolicy “Never” ,
+  indicating that the Initialization step is flaky and should be retried.
+- `Always` indicates that the container will be considered initialized the
+  moment the startup probe indicates it is started and will be restarted if it
+  exits for any reason until the pod terminates.
+- (If the initContainer's restartPolicy is unset, the pod's restartPolicy is
+  used to determine the Init container restart policy. For Always, OnFailure is
+  used, for OnFailure - OnFailure, for Never - Never).
+
+Since Init container and regular containers share the schema, restartPolicy
+field will be available for regular containers as well. From scenarios
+perspective, the first set of scenarios would be for containers that can only
+override the restartPolicy to “higher” value. From Never to OnFailure or Always,
+from OnFailure to Always. An override from Always to Never will not be
+allowed.This will eliminate situations when kubelet restarts and cannot decide
+whether the container overridden value to Never will need to be restarted or not
+for the Pod with the restart policy Always. It is also likely that we will
+require at least one container to have a restartPolicy matching the Pod
+restartPolicy. We can re-evaluate this in future if scenarios will arise.
+
+Note, by implementing these scenarios, the meaning of restartPolicy:Always for
+containers on a Pod with the restartPolicy:Never, will not exactly be the
+meaning of the restartPolicy:Always for the Pod with the restartPolicy:Always.
+For the first case, containers will not block the Pod termination. For the
+second, they will. In other words, the container with restartPolicy:Always will
+effectively become a sidecar container similar to Init container with the same
+field value.
+
+Additional considerations can be found in the older document that was a previous
+attempt to implement the sidecar containers:
+https://docs.google.com/document/d/1gX_SOZXCNMcIe9d8CkekiT0GU9htH_my8OB7HSm3fM8/edit#heading=h.71o8fcesvgba.
+
+### Naming
+
+#### Collection name
+
+For this KEP it is important to have sidecar containers be defined among other
+Init containers to be able to express the initialization order of containers.
+The name `initContainers` is not a good fit for sidecar containers as they
+typically do more than initialization. The better name would be “infrastructure”
+containers. The current idea is to implement sidecars as a part of init
+containers and if this introduces too much trouble, the new collection name may
+replace the old collection name in future. 
```

Comment:
The advantages of keeping them in initContainers are:

* The name initContainers implies something outside of normal operation of the pod
* More separate sections can lead to less clarity (users have to understand how they relate)
* initContainers have order, and we can reuse the ordering semantics to say that sidecar containers are started in FIFO order, and shut down in LIFO order, and if we want to (in the future) come up with parallel shutdown, we can do so without breaking existing users.

Reply:
the idea here is to replace the entire collection - deprecate Init and replace it with Infrastructure.

Reply:
polished text a little bit